pkcs11/pkcs11.go

```diff
@@ -1687,3 +1706,169 @@ func (s *Slot) generateECDSA(o keyOptions) (crypto.PrivateKey, error) {
+	}
+	r.hash = &hash
+	return r, nil
+}
+
+func (r *rsaPrivateKey) getHash() *crypto.Hash {
+	if r.hash != nil {
+		return r.hash
+	}
+	// Initialized to SHA256 if no hash specified
+	hash := crypto.SHA256
+	return &hash
+}
+
+func WithPublicKeyHandle(privKey crypto.PrivateKey, o Object) crypto.PrivateKey {
+	r := privKey.(*rsaPrivateKey)
+	r.pubH = o.o
+	return privKey
+}
+
+func (r *rsaPrivateKey) encryptRSA(data []byte) ([]byte, error) {
+	hash := r.getHash()
+	cParam, err := makeCParamRSAOAEP(hash)
+	if err != nil {
+		return nil, err
+	}
+	defer C.free(unsafe.Pointer(cParam))
+
+	cDataBytes := toCBytes(data)
+
+	m := makeMechanism(C.CKM_RSA_PKCS_OAEP, C.CK_VOID_PTR(cParam), C.sizeof_CK_RSA_PKCS_OAEP_PARAMS)
+
+	rv := C.ck_encrypt_init(r.o.fl, r.o.h, &m, r.pubH)
+	if err := isOk("C_EncryptInit", rv); err != nil {
+		return nil, err
+	}
+
+	var cCipherLen C.CK_ULONG
+
+	// First call is used to determine maximum length of encrypted data (PKCS #11 Specifications Section 5.2)
+	rv = C.ck_encrypt(r.o.fl, r.o.h, &cDataBytes[0], C.CK_ULONG(len(cDataBytes)), nil, &cCipherLen)
+	if err := isOk("C_Encrypt", rv); err != nil {
+		return nil, err
+	}
+
+	cCipher := make([]C.CK_BYTE, cCipherLen)
+
+	rv = C.ck_encrypt(r.o.fl, r.o.h, &cDataBytes[0], C.CK_ULONG(len(cDataBytes)), &cCipher[0], &cCipherLen)
+	if err := isOk("C_Encrypt", rv); err != nil {
+		return nil, err
+	}
+
+	cipher := toBytes(cCipher)
+
+	return cipher, nil
+}
+
+func (r *rsaPrivateKey) decryptRSA(encryptedData []byte) ([]byte, error) {
+	hash := r.getHash()
+	cParam, err := makeCParamRSAOAEP(hash)
+	if err != nil {
+		return nil, err
+	}
+	defer C.free(unsafe.Pointer(cParam))
+
+	cEncDataBytes := toCBytes(encryptedData)
+
+	m := makeMechanism(C.CKM_RSA_PKCS_OAEP, C.CK_VOID_PTR(cParam), C.sizeof_CK_RSA_PKCS_OAEP_PARAMS)
+
+	rv := C.ck_decrypt_init(r.o.fl, r.o.h, &m, r.o.o)
+	if err := isOk("C_DecryptInit", rv); err != nil {
+		return nil, err
+	}
+
+	var cDecryptedLen C.CK_ULONG
+
+	// First call is used to determine length necessary to hold decrypted data
+	rv = C.ck_decrypt(r.o.fl, r.o.h, &cEncDataBytes[0], C.CK_ULONG(len(cEncDataBytes)), nil, &cDecryptedLen)
+	if err := isOk("C_Decrypt", rv); err != nil {
+		return nil, err
+	}
+
+	cDecrypted := make([]C.CK_BYTE, cDecryptedLen)
+
+	rv = C.ck_decrypt(r.o.fl, r.o.h, &cEncDataBytes[0], C.CK_ULONG(len(cEncDataBytes)), &cDecrypted[0], &cDecryptedLen)
+	if err := isOk("C_Decrypt", rv); err != nil {
+		return nil, err
+	}
+
+	decrypted := toBytes(cDecrypted)
+
+	return decrypted, nil
+}
+
+func toBytes(data []C.CK_BYTE) []byte {
+	goBytes := make([]byte, len(data))
+	for i, b := range data {
+		goBytes[i] = byte(b)
+	}
+	return goBytes
+}
+
+func toCBytes(data []byte) []C.CK_BYTE {
+	cBytes := make([]C.CK_BYTE, len(data))
+	for i, b := range data {
+		cBytes[i] = C.CK_BYTE(b)
+	}
+	return cBytes
+}
+
+func makeMechanism(mech C.CK_MECHANISM_TYPE, paramPtr C.CK_VOID_PTR, paramLen int) C.CK_MECHANISM {
```

Comment:
nit: I don't find this more readable than:

```
 C.CK_MECHANISM{
	mechanism:      mech,
	pParameter:     paramPtr,
	ulParameterLen: C.CK_ULONG(paramLen),
}
```

Would you mind dropping this helper function?

Reply:
Okay, I will revert the change